### PR TITLE
Added Template based Project Wizard

### PR DESCRIPTION
**Pull request by keinhaar**
- Create projects based on templates.
- removed some unused code
- added a multi project template without example code
- added a multi project GWT project template with the example code from the existing GWT sample project, but splitted into 3 projects.

### CODE REVIEW

**Comment by protoism** [comment]
@keinhaar this PR looks great, and the implementation looks simply reasonable.
Before reviewing / merging, I'd like to know @niloc132 opinion about it.
Multi-module templates is definitely the right way to use GWT, classpath issues bite very often, but:

1) Is there any overlap with what the SDK does already?
2) The build is Eclipse only (no maven, no gradle), does this make sense? (I think so, honestly) - @keinhaar I don't find any pom.xml file, but I see something about maven nature. Am I rigth or not, that maven is not involved?

BTW the solution is expandable either by adding to the template dir, or defining an env variable. Not bad!

**Comment by keinhaar** [comment]
>     1. Is there any overlap with what the SDK does already?
It just uses modified versions of the files that are created by the sdk.
>     2. The build is Eclipse only (no maven, no gradle), does this make sense? (I think so, honestly) - @keinhaar I don't find any pom.xml file, but I see something about maven nature. Am I rigth or not, that maven is not involved?
maven based example could be added easily by just providing a template with pom files. I did not create more templates, because i did not know if this would be accepted.
> BTW the solution is expandable either by adding to the template dir, or defining an env variable. Not bad!
The env variable makes it possible to create and use your own custom templates. Thats what i personally would like to use, because my applications often have 5 or even 6 projects which depend on each other.

**Comment by FrankHossfeld** [comment]
@keinhaar Thanks for contributing.
I would suggest to provide only the multi module templates. GWT is moving to a more serverless approach. AFAIK the webappcreator inside GWT is deprecated for that reason (only one single module). And, as @protoism already mention, a Maven module artifact using the gwt-maven-plugin would be nice. 
Just an idea: and a template project using Spring Bott on the server side would be fine. ;-)

**Comment by FrankHossfeld** [comment]
> It just uses modified versions of the files that are created by the sdk

Just use the archetype creators for generating the template files:

- https://github.com/tbroyer/gwt-maven-archetypes
- https://github.com/NaluKit/gwt-maven-springboot-archetype (Spring Boot)

These are the suggested way creating a new project: https://www.gwtproject.org/gettingstarted-v2.html

.

**Comment by keinhaar** [comment]
@niloc132 Could you please have a look on this pull request, as @protoism  requested?

File: plugins/com.gwtplugins.gdt.eclipse.suite/src/com/google/gdt/eclipse/suite/wizards/NewWebAppTemplateProjectWizard.java [hunk 1]
```diff
@@ -0,0 +1,130 @@
+/*******************************************************************************
+ * Copyright 2011 Google Inc. All Rights Reserved.
```
Comment (niloc132) [inline plugins/com.gwtplugins.gdt.eclipse.suite/src/com/google/gdt/eclipse/suite/wizards/NewWebAppTemplateProjectWizard.java:2]: new files, should have updated copyright date, and owner

File: plugins/com.gwtplugins.gdt.eclipse.suite/src/com/google/gdt/eclipse/suite/wizards/ProjectTemplate.java [hunk 2]
```diff
@@ -0,0 +1,171 @@
+/**
+ *
```
Comment (niloc132) [inline plugins/com.gwtplugins.gdt.eclipse.suite/src/com/google/gdt/eclipse/suite/wizards/ProjectTemplate.java:2]: should be populated, or removed?

File: plugins/com.gwtplugins.gwt.eclipse.wtp/src/com/google/gwt/eclipse/wtp/projects/WtpFacetCreatorParicipant.java [hunk 3]
```diff
@@ -19,14 +19,14 @@
...
+  private IJavaProject[] javaProjects;
   private IFacetedProjectListener projectFacetListener;
  
   @Override
   public void updateWebAppProjectCreator(IWebAppProjectCreator webAppProjectCreator) {
     // Just in case exit early
-    javaProject = webAppProjectCreator.getCreatedJavaProject();
-    if (javaProject == null) {
+    javaProjects = webAppProjectCreator.getCreatedJavaProjects();
+    if (javaProjects == null) {
```
Comment (niloc132) [inline plugins/com.gwtplugins.gwt.eclipse.wtp/src/com/google/gwt/eclipse/wtp/projects/WtpFacetCreatorParicipant.java:29]: it doesnt appear that this should never return null - would it make more sense to test if empty? also, what do you think about `List<IJavaProject>`?

File: pom.xml [hunk 4]
```diff
@@ -11,7 +11,7 @@
   <properties>
     <tycho.version>3.0.1</tycho.version>
     <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
-    <eclipse.target>2022-09</eclipse.target> <!-- TODO change to neon -->
+    <eclipse.target>2023-09</eclipse.target> <!-- TODO change to neon -->
```
Comment (niloc132) [inline pom.xml:14]: probably can remove that TODO as long as you're updating the line...

File: plugins/com.gwtplugins.gdt.eclipse.suite/project_templates/client_server_shared/client/src/_PACKAGENAME_/ui.gwt.xml [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+  <inherits name='${PACKAGENAME}.shared'/>
+
+  <!-- Specify the app entry point class.                         -->
+  <entry-point class='${PACKAGENAME}.ui.Main'/>
+
+  <!-- Specify the paths for translatable code                    -->
+  <source path='ui'/>
+
+  <!-- allow Super Dev Mode -->
+  <add-linker name="xsiframe"/>
```
Comment (niloc132) [inline plugins/com.gwtplugins.gdt.eclipse.suite/project_templates/client_server_shared/client/src/_PACKAGENAME_/ui.gwt.xml:30]: No longer needed since GWT 2.7, almost 10 years https://github.com/gwtproject/gwt/commit/22865ad27d54e2eb03a017feefc482c59afbc7fe

File: plugins/com.gwtplugins.gdt.eclipse.suite/src/com/google/gdt/eclipse/suite/wizards/NewWebAppTemplateProjectWizardPage.java [hunk 6]
```diff
@@ -0,0 +1,532 @@
...
+    try {
+      templates = ProjectTemplate.getTemplates();
+      Collections.sort(templates);
+      for(int i=0;i<templates.size();i++)
+      {
+        ProjectTemplate templ = templates.get(i);
+        templateCombo.add(templ.getName());
+      }
+    } catch (IOException e1) {
+      e1.printStackTrace();
```
Comment (niloc132) [inline plugins/com.gwtplugins.gdt.eclipse.suite/src/com/google/gdt/eclipse/suite/wizards/NewWebAppTemplateProjectWizardPage.java:187]: probably should throw or return early so we halt the process? it seems bad to stop with some of the templates applied but not all, and no negative feedback (except this trace in the logs)

File: plugins/com.gwtplugins.gdt.eclipse.suite/src/com/google/gdt/eclipse/suite/wizards/WebAppTemplateProjectCreator.java [hunk 7]
```diff
@@ -0,0 +1,435 @@
...
+  private void replacePackageNames(IProject iProject, File dir) throws IOException {
+    Stack<File> files = new Stack<>();
+    files.add(dir);
+    while(files.isEmpty() == false)
+    {
+      File file = files.pop();
+      if(file.getName().contains("_PACKAGENAME_"))
+      {
+        String name = file.getName();
+        name = name.replace("_PACKAGENAME_", packageBaseName.replace('.', File.separatorChar));
```
Comment (niloc132) [inline plugins/com.gwtplugins.gdt.eclipse.suite/src/com/google/gdt/eclipse/suite/wizards/WebAppTemplateProjectCreator.java:215]: Probably a stretch, but should this perhaps be replaceAll? This seems like the least likely one to change to replaceAll, so I'll make the case here and let you decide for the others: if templates ever end up having super-source in them, then there will be a case of

src/main/java/com/company/project/super/com/company/project/SomeReplacedClass.java

so we _could_ want to handle this better?

File: plugins/com.gwtplugins.gdt.eclipse.suite/src/com/google/gdt/eclipse/suite/wizards/WebAppTemplateProjectCreator.java [hunk 8]
```diff
@@ -0,0 +1,435 @@
...
+        file = newFile;
+      }
+      if(file.isDirectory())
+      {
+        File[] cf = file.listFiles();
+        files.addAll(Arrays.asList(cf));
+      }
+      else
+      {
+        replaceInFile(file.toString(), "\\$\\{PACKAGENAME\\}", packageBaseName);
```
Comment (niloc132) [inline plugins/com.gwtplugins.gdt.eclipse.suite/src/com/google/gdt/eclipse/suite/wizards/WebAppTemplateProjectCreator.java:237]: I'm not a huge fan of supporting multiple template patterns in the same file, but maybe we can be consistent here, and always use `_` for paths, and `${}`s for strings?

That said, replaceInFile never actually accepts a pattern - why not change the string to take a literal, and use `String pattern = Pattern.quote(findMe)` to rewrite? then we never need to worry about mixing tokens

File: plugins/com.gwtplugins.gdt.eclipse.suite/src/com/google/gdt/eclipse/suite/wizards/WebAppTemplateProjectCreator.java [hunk 9]
```diff
@@ -0,0 +1,435 @@
...
+   */
+  public static void replaceInFile(String file, String pattern, String replacement) throws IOException {
+    Charset cset = Charset.defaultCharset();
+    java.nio.file.Path path = Paths.get(file);
+    if (Files.exists(path)) {
+      if (Files.size(path) > 1000000) {
+        throw new IOException("File to large");
+      }
+      byte[] data = Files.readAllBytes(path);
+      String str = new String(data, cset);
```
Comment (niloc132) [inline plugins/com.gwtplugins.gdt.eclipse.suite/src/com/google/gdt/eclipse/suite/wizards/WebAppTemplateProjectCreator.java:258]: Are you sure we want the default charset here, rather than just assuming utf8 everywhere?

File: plugins/com.gwtplugins.gdt.eclipse.suite/src/com/google/gdt/eclipse/suite/wizards/WebAppTemplateProjectCreator.java [hunk 10]
```diff
@@ -0,0 +1,435 @@
...
+      String newStr = str.replaceAll(pattern, replacement);
+      BufferedWriter writer = Files.newBufferedWriter(path, cset);
+      writer.write(newStr);
+      writer.close();
```
Comment (niloc132) [inline plugins/com.gwtplugins.gdt.eclipse.suite/src/com/google/gdt/eclipse/suite/wizards/WebAppTemplateProjectCreator.java:262]: try-with-resources, otherwise if the write fails this might not get closed (and then you've got a broken project, and windows won't let it be deleted or the like)

(goes away if you use Files.write())

File: plugins/com.gwtplugins.gdt.eclipse.core/src/com/google/gdt/eclipse/core/resources/ProjectResources.java [hunk 11]
```diff
@@ -128,13 +128,13 @@ public static List<File> findFilesInDir(File directory, FilenameFilter filter) {
...
    * code on that source and write it back to disk.
-   * 
+   *
    * @param file
    * @throws CoreException
    */
   public static void reformatJavaSource(File file) throws CoreException {
     try {
-      String generatedSource = textFromFile(file);
+      String generatedSource = Files.readString(file.toPath(), Charset.defaultCharset());
```
Comment (niloc132) [inline plugins/com.gwtplugins.gdt.eclipse.core/src/com/google/gdt/eclipse/core/resources/ProjectResources.java:137]: consider if we really want the default charset here, or just assume utf8?

File: plugins/com.gwtplugins.gdt.eclipse.suite/src/com/google/gdt/eclipse/suite/wizards/WebAppTemplateProjectCreator.java [hunk 12]
```diff
@@ -0,0 +1,435 @@
...
+    java.nio.file.Path path = Paths.get(file);
+    if (Files.exists(path)) {
+      if (Files.size(path) > 1000000) {
+        throw new IOException("File to large");
+      }
+      byte[] data = Files.readAllBytes(path);
+      String str = new String(data, cset);
+      String newStr = str.replaceAll(pattern, replacement);
+      BufferedWriter writer = Files.newBufferedWriter(path, cset);
+      writer.write(newStr);
```
Comment (niloc132) [inline plugins/com.gwtplugins.gdt.eclipse.suite/src/com/google/gdt/eclipse/suite/wizards/WebAppTemplateProjectCreator.java:261]: if we've got the full string in memory... why buffer? why not just use Files.write(...)?

File: plugins/com.gwtplugins.gdt.eclipse.suite/src/com/google/gdt/eclipse/suite/wizards/WebAppTemplateProjectCreator.java [hunk 13]
```diff
@@ -0,0 +1,435 @@
...
+    files.add(dir);
+    while(files.isEmpty() == false)
+    {
+      File file = files.pop();
+      if(file.getName().contains("_PACKAGENAME_"))
+      {
+        String name = file.getName();
+        name = name.replace("_PACKAGENAME_", packageBaseName.replace('.', File.separatorChar));
+        File newFile = new File(file.getParentFile(), name);
+        newFile.getParentFile().mkdirs();
```
Comment (niloc132) [inline plugins/com.gwtplugins.gdt.eclipse.suite/src/com/google/gdt/eclipse/suite/wizards/WebAppTemplateProjectCreator.java:217]: consider using java.nio.Files.createDirectories(..) here - the Files class will throw instead of silently fail (or letting you ignore the return value, as you're doing here).

**Review by niloc132 — COMMENTED** [review]
A lot of this is pretty eclipse-specific, I'm not sure how helpful I can be. Here's a quick review, but I think it would make more sense for someone who knows the project more deeply to dig in than me.

File: plugins/com.gwtplugins.gdt.eclipse.core/src/com/google/gdt/eclipse/core/resources/ProjectResources.java (hunk 11 above)
Comment (keinhaar) [reply plugins/com.gwtplugins.gdt.eclipse.core/src/com/google/gdt/eclipse/core/resources/ProjectResources.java:137]: This has been default charset in the previous implementation. Didn't want to change that. Just simplified the Code by using an existing JRE method.

File: plugins/com.gwtplugins.gdt.eclipse.suite/src/com/google/gdt/eclipse/suite/wizards/NewWebAppTemplateProjectWizard.java (hunk 1 above)
Comment (keinhaar) [reply plugins/com.gwtplugins.gdt.eclipse.suite/src/com/google/gdt/eclipse/suite/wizards/NewWebAppTemplateProjectWizard.java]: Fixed this

File: plugins/com.gwtplugins.gdt.eclipse.suite/src/com/google/gdt/eclipse/suite/wizards/ProjectTemplate.java (hunk 2 above)
Comment (keinhaar) [reply plugins/com.gwtplugins.gdt.eclipse.suite/src/com/google/gdt/eclipse/suite/wizards/ProjectTemplate.java:8]: Fixed this

File: plugins/com.gwtplugins.gwt.eclipse.wtp/src/com/google/gwt/eclipse/wtp/projects/WtpFacetCreatorParicipant.java (hunk 3 above)
Comment (keinhaar) [reply plugins/com.gwtplugins.gwt.eclipse.wtp/src/com/google/gwt/eclipse/wtp/projects/WtpFacetCreatorParicipant.java:31]: changed this to list

File: pom.xml (hunk 4 above)
Comment (keinhaar) [reply pom.xml]: removed todo

File: plugins/com.gwtplugins.gdt.eclipse.suite/project_templates/client_server_shared/client/src/_PACKAGENAME_/ui.gwt.xml (hunk 5 above)
Comment (keinhaar) [reply plugins/com.gwtplugins.gdt.eclipse.suite/project_templates/client_server_shared/client/src/_PACKAGENAME_/ui.gwt.xml]: removed the xsiframe

File: plugins/com.gwtplugins.gdt.eclipse.suite/src/com/google/gdt/eclipse/suite/wizards/WebAppTemplateProjectCreator.java (hunk 9 above)
Comment (keinhaar) [reply plugins/com.gwtplugins.gdt.eclipse.suite/src/com/google/gdt/eclipse/suite/wizards/WebAppTemplateProjectCreator.java:268]: changed to UTF-8

File: plugins/com.gwtplugins.gdt.eclipse.suite/src/com/google/gdt/eclipse/suite/wizards/WebAppTemplateProjectCreator.java (hunk 10 above)
Comment (keinhaar) [reply plugins/com.gwtplugins.gdt.eclipse.suite/src/com/google/gdt/eclipse/suite/wizards/WebAppTemplateProjectCreator.java]: changed to Files.writeString()

File: plugins/com.gwtplugins.gdt.eclipse.suite/src/com/google/gdt/eclipse/suite/wizards/WebAppTemplateProjectCreator.java (hunk 12 above)
Comment (keinhaar) [reply plugins/com.gwtplugins.gdt.eclipse.suite/src/com/google/gdt/eclipse/suite/wizards/WebAppTemplateProjectCreator.java]: see comment above

File: plugins/com.gwtplugins.gdt.eclipse.suite/src/com/google/gdt/eclipse/suite/wizards/WebAppTemplateProjectCreator.java (hunk 13 above)
Comment (keinhaar) [reply plugins/com.gwtplugins.gdt.eclipse.suite/src/com/google/gdt/eclipse/suite/wizards/WebAppTemplateProjectCreator.java]: changed to createDirectories

**Comment by keinhaar** [comment]
> > It just uses modified versions of the files that are created by the sdk
> 
> Just use the archetype creators for generating the template files:
> 
>     * https://github.com/tbroyer/gwt-maven-archetypes (Tomcat)
> 
>     * https://github.com/NaluKit/gwt-maven-springboot-archetype (Spring Boot)
> 
> 
> These are the suggested way creating a new project: https://www.gwtproject.org/gettingstarted-v2.html

I've created a maven based template using the instructions on the gettingstarted-v2 page. It runs just fine if used with maven, but there are some problems running it from eclipse:
1. It uses the jakarta enabled version of gwt-servlet, which is currently not supported by the plugin (maybe if you rename it manually?)
2. The module.gwt.xml specifies an empty source path, which will not be accepted when running from eclipse
3. The Eclipse Plugin can't create a valid launch command. I was able to create it manually and it works, but then maven will no longer work because of the changed source path.

To make it work from eclipse the following steps are required:
- change **`<source path=""/>`** to **`<source path="MAINPACKAGE"/>`**
- create a launcher manually with these arguments:
`
-remoteUI "${gwt_remote_ui_server_port}:${unique_id}" -startupUrl index.html -logLevel INFO -codeServerPort 9997 -port 8888 -war /YOURPATH/PROJECTNAME/PROJECTNAME-server/src/main/webapp module module
`

I'm in doubt if i should add it in this state.
Because i don't use maven, some maven user may decide this. @protoism, @FrankHossfeld? What do you think?

**Comment by keinhaar** [comment]
@protoism 
I've fixed all the things that collin suggested.

**Review by protoism — APPROVED** [review]
I'm not sure I like the idea of having the source inside the plugin.
And I really wonder if having GWT projects without maven or gradle make really sense.
Nonetheless, making it clear the best practice is splitting is so important that we can't be fussy on this :)
The code looks ok.
Kudos, and thanks.

**Comment by protoism** [comment]
PS. thanks @niloc132 for the careful review
